File: cs/src/AutoDiff/Reification.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
	[Serializable]
	 /// <summary>
    /// Represents a reified constraint.
    /// </summary>
    public class Reification : Term
    {
       
        public Reification(Term condition, double min,double max)
        {
            Condition = condition;
			NegatedCondition = condition.Negate();
            MinVal = min;
			MaxVal = max;
        }
		/// <summary>
		/// Gets or sets the constraint to reify
		/// </summary>
		/// <value>
		/// The constraint.
		/// </value>
        public Term Condition { get; private set; }
		/// <summary>
		/// The constraint's negation. Set automatically.
		/// </summary>
		/// <value>
		/// The negated constraint.
		/// </value>
		public Term NegatedCondition { get; private set; }
		/// <summary>
		/// Gets or sets the minimum.
		/// </summary>
		/// <value>
		/// The value representing a violated constraint.
		/// </value>
	    public double MinVal { get; private set; }
		/// <summary>
		/// Gets or sets the maximum.
		/// </summary>
		/// <value>
		/// The value representing a satisfied constraint.
		/// </value>
		public double MaxVal { get; private set; }

        /// <summary>
        /// Accepts a term visitor
        /// </summary>
        /// <param name="visitor">The term visitor to accept</param>
        public override void Accept(ITermVisitor visitor)
        {
            visitor.Visit(this);
        }

        /// <summary>
        /// Accepts a term visitor with a generic result
        /// </summary>
        /// <typeparam name="TResult">The type of the result from the visitor's function</typeparam>
        /// <param name="visitor">The visitor to accept</param>
        /// <returns>
        /// The result from the visitor's visit function.
        /// </returns>
        public override TResult Accept<TResult>(ITermVisitor<TResult> visitor)
        {
            return visitor.Visit(this);
        }
		
		
		public override string ToString()
		{
			return string.Format("Discretizer( {0}, {1}, {2} )",Condition,Min, Max);
		}
		
		public override Term AggregateConstants()
		{
			return this;
		}
		public override Term Derivative(Variable v)
		{
			throw new Exception("Symbolic Derivation of Discretizer not supported.");
		}

    }
}

File: java/src/main/java/AutoDiff/Reification.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:34
//

package AutoDiff;

/**
* Represents a reified constraint.
*/
public class Reification  extends Term 
{
    public Reification(Term condition, double min, double max) throws Exception {
        setCondition(condition);
        setNegatedCondition(condition.negate());
        setMinVal(min);
        setMaxVal(max);
    }

    /**
    * Gets or sets the constraint to reify
    * 
    * The constraint.
    */
    private Term Condition;
    public Term getCondition() {
        return Condition;
    }

    public void setCondition(Term value) {
        Condition = value;
    }

    /**
    * The constraint's negation. Set automatically.
    * 
    * The negated constraint.
    */
    private Term NegatedCondition;
    public Term getNegatedCondition() {
        return NegatedCondition;
    }

    public void setNegatedCondition(Term value) {
        NegatedCondition = value;
    }

    /**
    * Gets or sets the minimum.
    * 
    * The value representing a violated constraint.
    */
    private double MinVal;
    public double getMinVal() {
        return MinVal;
    }

    public void setMinVal(double value) {
        MinVal = value;
    }

    /**
    * Gets or sets the maximum.
    * 
    * The value representing a satisfied constraint.
    */
    private double MaxVal;
    public double getMaxVal() {
        return MaxVal;
    }

    public void setMaxVal(double value) {
        MaxVal = value;
    }
//
//    /**
//    * Accepts a term visitor
//    *
//    *  @param visitor The term visitor to accept
//    */
//    public void accept(ITermVisitor visitor) throws Exception {
//        visitor.visit(this);
//    }

    /**
    * Accepts a term visitor with a generic result
    * The type of the result from the visitor's function
    *  @param visitor The visitor to accept
    *  @return 
    * The result from the visitor's visit function.
    */
    public <TResult>TResult accept(ITermVisitor<TResult> visitor) throws Exception {
        return visitor.visit(this);
    }

    public String toString() {
        try
        {
            return String.format("Discretizer( %s, %f, %f )", getCondition(), Min, Max);
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    public Term aggregateConstants() throws Exception {
        return this;
    }

    public Term derivative(Variable v) throws Exception {
        throw new Exception("Symbolic Derivation of Discretizer not supported.");
    }

}


